Add custom NovixBottomSheetState & Refactor filtering sheet

yusufnasserdev opened the pull request:
# What does this PR do?

This pull request refactors the filter functionality within the search feature to improve state management and UI consistency.

The core of this update is moving the filter bottom sheet's state management from the SearchScreen composable into the SearchViewModel. This centralizes the logic, making the UI more declarative and less prone to state-related bugs.

Key Changes:
- A new custom `NovixBottomSheetState` component has been created as a wrapper around Material 3's `ModalBottomSheetState`. It comes with a `rememberNovixModalBottomSheetState` function that defaults to skipping the partially-expanded state, ensuring a consistent bottom sheet experience throughout the app.

- Centralized State in `SearchViewModel`: The `SearchViewModel` now manages the visibility and data for the filter bottom sheet.  New methods (onFilterClick, onFilterSheetDismiss) have been added to the SearchInteractions interface to handle user actions.

- Refactored SearchScreen and FilterBottomSheet:

  - The SearchScreen no longer uses local state to control the filter sheet. Instead, it observes the state from the SearchViewModel.

  - The FilterBottomSheet has been streamlined to accept a FilterState data class and SearchInteractions, making it a more reusable and declarative component that leverages the new NovixBottomSheet.

## Type of Change
- 🐛 Bug fix
- ✨ New feature
- 🔧 Refactoring

## Changes Made
- Compose UI

## Testing
- Tested locally

## Checklist
- [x] Self-reviewed
- [x] Ready for review

yusufnasserdev committed:
feat: create `NovixBottomSheetState` and `rememberNovixModalBottomSheetState`

This commit introduces a custom state management function, `rememberNovixModalBottomSheetState` as a wrapper for m3 `rememberModalBottomSheetState`.

The `NovixBottomSheet` wraps the Material 3 `ModalBottomSheet` and provides a more opinionated API by:
- Renaming the component from `ModalBottomSheet` to `NovixBottomSheet`.
- Removing the `skipPartiallyExpanded` parameter, as it's now handled by the custom state.
- Exposing a `NovixModalBottomSheetState` for better control over the sheet's visibility and state.

The `rememberNovixModalBottomSheetState` function:
- Creates and remembers a `NovixModalBottomSheetState`.
- Allows customization of the `skipPartiallyExpanded` behavior (defaults to true).

yusufnasserdev committed:
feat: add filter callbacks to search viewmodel & interactions

- Adding `onFilterClick` and `onFilterSheetDismiss` methods to the `SearchInteractions` interface.
- Implementing these methods in `SearchViewModel` to update the state and show/hide the filter bottom sheet.

yusufnasserdev committed:
refactor: update FilterBottomSheet handling in SearchScreen

Key changes:
- Removed the local `showFilterBottomSheet` state in `SearchScreen` and moved its management to `SearchViewModel` through `SearchUiState`.
- Connected `FilterBottomSheet`'s visibility and data to `SearchUiState`.
- Passed `SearchViewModel` as `filterInteractions` to `FilterBottomSheet`.
- Updated the filter button click to call `viewModel::onFilterClick`.
- Adjusted animation and spacing for the filter button and search field.

yusufnasserdev committed:
refactor: improve filter bottom sheet

This commit refactors the `FilterBottomSheet` composable to:

- Use a `FilterState` data class to manage its state, making it more declarative.
- Utilize `NovixBottomSheet` for a consistent look and feel with other bottom sheets in the app.
- Improve state management for visibility and dismissal by using `LaunchedEffect` and coroutine scope.
- Enhance the interaction with the `SearchViewModel` through a dedicated `SearchInteractions` interface.
- Streamline the content of the bottom sheet by passing the `FilterState` and `SearchInteractions` directly.
- Ensure the close icon is correctly positioned using `Arrangement.SpaceBetween`.

london-bot[bot] committed:
Branch was auto-updated.

ahmeddsamii commented on designSystem/src/main/java/com/london/designsystem/component/NovixBottomSheet.kt:
```diff
@@ -4,35 +4,55 @@ import androidx.compose.foundation.layout.ColumnScope
+    suspend fun show() = sheetState.show()
+    suspend fun hide() = sheetState.hide()
+    val isVisible: Boolean get() = sheetState.isVisible
+    val currentValue: SheetValue get() = sheetState.currentValue
```
this is not used

yusufnasserdev replied:
could be needed, keep in mind that this is a wrapper for an existing material 3 component.

ahmeddsamii replied:
okay then you can pass them with its defuault value

ahmeddsamii commented on designSystem/src/main/java/com/london/designsystem/component/NovixBottomSheet.kt:
```diff
@@ -4,35 +4,55 @@ import androidx.compose.foundation.layout.ColumnScope
+    val isVisible: Boolean get() = sheetState.isVisible
+    val currentValue: SheetValue get() = sheetState.currentValue
+    val targetValue: SheetValue get() = sheetState.targetValue
+    val hasExpanded: Boolean get() = sheetState.hasExpandedState
```
not used

yusufnasserdev replied:
could be needed, keep in mind that this is a wrapper for an existing material 3 component.

ahmeddsamii replied:
okay then you can pass them with its defuault value

yusufnasserdev replied:
I don't get you, there value is provided by the wrapped sheetState.

ahmeddsamii commented on designSystem/src/main/java/com/london/designsystem/component/NovixBottomSheet.kt:
```diff
@@ -4,35 +4,55 @@ import androidx.compose.foundation.layout.ColumnScope
+    val currentValue: SheetValue get() = sheetState.currentValue
+    val targetValue: SheetValue get() = sheetState.targetValue
+    val hasExpanded: Boolean get() = sheetState.hasExpandedState
+    val hasPartiallyExpanded: Boolean get() = sheetState.hasPartiallyExpandedState
```
not used

yusufnasserdev replied:
could be needed, keep in mind that this is a wrapper for an existing material 3 component.

ahmeddsamii replied:
okay then you can pass them with its defuault value

yusufnasserdev replied:
I don't get you, there value is provided by the wrapped sheetState.

ahmeddsamii commented on presentation/src/main/java/com/london/presentation/composables/filterbottomsheet/FilterBottomSheet.kt:
```diff
@@ -30,95 +28,97 @@ import androidx.compose.ui.unit.dp
 import kotlinx.coroutines.launch
 import org.koin.androidx.compose.koinViewModel
 
+data class FilterState(
```
the file name is FilterBottomSheet so i think this data class should be in another file because when i open this file i expect to see only filterBottomSheet

yusufnasserdev replied:
I agree it shouldn't be here, it should be in the SearchUiState instead as a part of it, however, I'm leaving this for a later refactor, this pr is more concerned with exposing the sheetstate and having it work as it did before removing m3 from presentation.

ahmeddsamii replied:
i mean why we should leave it for later while you can simply move it in another class to forget about it and we will not back again

yusufnasserdev replied:
Becasue refactoring the SearchUiState could have other implications, and if I started there this PR would be concerned with a whole lot more that it is.

ahmeddsamii commented on presentation/src/main/java/com/london/presentation/composables/filterbottomsheet/FilterBottomSheet.kt:
```diff
@@ -30,95 +28,97 @@ import androidx.compose.ui.unit.dp
+    val releaseYearRange: ClosedFloatingPointRange<Float>
+)
+
 @SuppressLint("ConfigurationScreenWidthHeight")
```
you can remove this and for LocalConfiguration.current.screenHeightDp -> u can use LocalWindowInfo.current.containerSize.height.dp instead and it will not give you a warning and you don't need this SuppressLint anymore or you can use local density which ahmed mones recommend

yusufnasserdev replied:
This is not a change I introduced, but I'll check this, thanks for the note.

ahmeddsamii commented on presentation/src/main/java/com/london/presentation/composables/filterbottomsheet/FilterBottomSheet.kt:
```diff
@@ -30,95 +28,97 @@ import androidx.compose.ui.unit.dp
-    onRatingChanged: (Int) -> Unit,
+    filterInteractions: SearchInteractions,
+    filterUiState: FilterState,
+    onCloseClicked: () -> Unit,
```
why we need this click while we have interactions now? should't this click to be handled in interactions?

ahmeddsamii commented on presentation/src/main/java/com/london/presentation/composables/filterbottomsheet/FilterBottomSheet.kt:
```diff
@@ -30,95 +28,97 @@ import androidx.compose.ui.unit.dp
+    filterInteractions: SearchInteractions,
+    filterUiState: FilterState,
+    onCloseClicked: () -> Unit,
+    onApplyClicked: () -> Unit,
```
same here

yusufnasserdev replied:
Interactions are defined by the viewmodel, however the onApplyClicked uses the sheetState which is only available for this component.

```kotlin
onApplyClicked = {
    filterInteractions.onApplyFilter(
        filterState.selectedGenres,
        filterState.minimumImdbRating,
        filterState.releaseYearRange
    )
    coroutineScope.launch {
        sheetState.hide()
    }.invokeOnCompletion {
        if (!sheetState.isVisible) {
            filterInteractions.onFilterSheetDismiss()
        }
    }
}
```

Note that interactions are still passed for other functionality, however this and `onClose` are both needed to be this way as explained.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

marwanqashwa commented on designSystem/src/main/java/com/london/designsystem/component/NovixBottomSheet.kt:
```diff
@@ -4,35 +4,55 @@ import androidx.compose.foundation.layout.ColumnScope
+}
+
+@Composable
+fun NovixBottomSheet(
```
Remove novix prefix from class and methods call material components with full path

yusufnasserdev replied:
Why remove the prefix? using the same name as the m3 one can call for confusion in the presentation or other modules, I want the caller to understand that this is a custom one.

marwanqashwa replied:
we dont't need it as we talk it's our components and we know that

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

yusufnasserdev committed:
refactor: rename NovixBottomSheet to ModalBottomSheet

The `NovixBottomSheetState` has been renamed to `SheetState` for clarity.

Additionally, the calculation for the maximum height of the filter bottom sheet now uses `LocalWindowInfo.current.containerSize.height` instead of `LocalConfiguration.current.screenHeightDp`.

yusufnasserdev committed:
refactor: rename `rememberNovixModalBottomSheetState` to `rememberModalBottomSheetState`

This change renames the custom `rememberNovixModalBottomSheetState` composable to `rememberModalBottomSheetState` to align with the standard Material3 naming convention. Additionally, unused imports were removed from `FilterBottomSheet.kt`.
